mls-rs/src/group/mod.rs

```diff
@@ -745,34 +755,57 @@ where
             }
         }
 
-        // Apply own update
+        // Apply own update or a Replace proposal replacing us
         let new_signer = None;
 
-        #[cfg(feature = "by_ref_proposal")]
+        #[cfg(any(feature = "by_ref_proposal", feature = "replace_proposal"))]
         let mut new_signer = new_signer;
 
-        #[cfg(feature = "by_ref_proposal")]
-        for p in &provisional_state.applied_proposals.updates {
-            if p.sender == Sender::Member(*self_index) {
-                let leaf_pk = &p.proposal.leaf_node.public_key;
+        #[cfg(any(feature = "by_ref_proposal", feature = "replace_proposal"))]
+        {
+            let updated_leaves = std::iter::empty();
 
-                // Update the leaf in the private tree if this is our update
-                #[cfg(feature = "std")]
-                let new_leaf_sk_and_signer = self.pending_updates.get(leaf_pk);
+            #[cfg(feature = "by_ref_proposal")]
+            let updated_leaves = updated_leaves.chain(
+                provisional_state
+                    .applied_proposals
+                    .update_senders
+                    .iter()
+                    .zip(provisional_state.applied_proposals.updates.iter())
+                    .filter(|(&i, _p)| i.0 == *self_index)
+                    .map(|(_i, p)| p.proposal.leaf_node.public_key.clone()),
+            );
 
-                #[cfg(not(feature = "std"))]
-                let new_leaf_sk_and_signer = self
-                    .pending_updates
+            #[cfg(feature = "replace_proposal")]
+            let updated_leaves = updated_leaves.chain(
+                provisional_state
+                    .applied_proposals
+                    .replaces
                     .iter()
-                    .find_map(|(pk, sk)| (pk == leaf_pk).then_some(sk));
+                    .filter(|p| p.proposal.to_replace.0 == *self_index)
+                    .map(|p| p.proposal.leaf_node.public_key.clone()),
+            );
+
+            let mut updated_leaves = updated_leaves;
+            let self_update = updated_leaves.next();
+
+            if updated_leaves.count() > 0 {
+                // XXX(RLB) This error code is wrong; might need a new value?  Or just not bother
+                // checking, assuming checks have been done upstream?  But then we might also need a
+                // new error code there.  In any case, we need to enforce that there are not multiple
+                // Update **or Replace** proposals for the same leaf.
+                return Err(MlsError::InvalidCommitSelfUpdate);
```

Comment:
Didn't we already made sure of that when validating the proposal list? So is this check redundant?